### PR TITLE
Update readme for 3.0.0

### PR DESCRIPTION
**Pull request by uschi2000**

### CODE REVIEW

**Comment by uschi2000** [comment]
TODO: deprecate WebApplicationException

**Review by markelliot — APPROVED** [review]

File: readme.md [hunk 1]
```diff
@@ -8,11 +8,11 @@ on [Feign](https://github.com/OpenFeign/feign) or [Retrofit](http://square.githu
 service definitions as a server. Refer to the [API Contract](#api-contract) section for details on the contract between
 clients and servers. This library requires Java 8.
 
-
 Core libraries:
 - jaxrs-clients: Clients for JAX-RS-defined service interfaces
 - retrofit2-clients: Clients for Retrofit-defined service interfaces
 - jersey-servers: Configuration library for Dropwizard/Jersey servers
+- [http-remoting-api](https://github.com/palantir/http-remoting-api): API classes for service configuration
```
Comment (gracew) [inline readme.md]: also service errors and tracing

File: readme.md [hunk 2]
```diff
@@ -123,27 +125,31 @@ try {
...
-MyService service = JaxRsClient.create(
-    MyService.class, "my user agent", config.getServices().get("myService"));
+ServiceConfigBlock config = readFromYaml("config.yml");
+ServiceConfigurationFactory factory = ServiceConfigurationFactory.of(config);
+MyService client = JaxRsClient.create(MyService.class, "my-agent", ClientConfigurations.of(factory.get("myService")));
 ```
 
 
-## ssl-config
+## keystores and ssl-config (http-remoting)
```
Comment (gracew) [inline readme.md]: http-remoting-api

File: readme.md [hunk 3]
```diff
@@ -227,28 +233,67 @@ Jackson `ObjectMapper` with `GuavaModule`, `ShimJdk7Module` (same as Jackson’s
...
-message, exception name, and optionally stacktrace of the exception. Both JaxRsClient and Retrofit2Client intercept
-non-successful HTTP responses and throw a `RemoteException` wrapping the deserialized server-side `SerializableError`.
-The error name of the `RemoteException` is defined by the service API and clients should switch&dispatch based on the
-error name. The `SerializableError` format is:
+Servers should use the `ServiceException` class to propagate application-specific errors to its callers. Typically, 
+services define its error types as follows:
+
+```
+public static final ErrorType DATASET_NOT_FOUND = 
+  ErrorType.create(ErrorType.Code.INVALID_ARGUMENT, "DatasetNotFound");
```
Comment (gracew) [inline readme.md]: should include namespace in example

File: readme.md [hunk 4]
```diff
@@ -227,28 +233,67 @@ Jackson `ObjectMapper` with `GuavaModule`, `ShimJdk7Module` (same as Jackson’s
...
+
 #### Error propagation
-The `HttpRemotingJerseyFeature` routine installs exception mappers for `IllegalArgumentException`,
-`NoContentException`, `RuntimeException` and `WebApplicationException`. The exception mapper sets the response media
-type to `application/json` and returns as response body a JSON representation of a `SerializableError` capturing the
-message, exception name, and optionally stacktrace of the exception. Both JaxRsClient and Retrofit2Client intercept
-non-successful HTTP responses and throw a `RemoteException` wrapping the deserialized server-side `SerializableError`.
-The error name of the `RemoteException` is defined by the service API and clients should switch&dispatch based on the
-error name. The `SerializableError` format is:
+Servers should use the `ServiceException` class to propagate application-specific errors to its callers. Typically,
```
Comment (gracew) [inline readme.md]: should link to safe-logging and mention that ServiceException implements SafeLoggable

File: readme.md [hunk 5]
```diff
@@ -227,28 +233,67 @@ Jackson `ObjectMapper` with `GuavaModule`, `ShimJdk7Module` (same as Jackson’s
...
+        // Note that only SafeArg parameters are sent to the caller in the resulting SerializableError.
+        throw new ServiceException(
+                DATASET_NOT_FOUND, 
+                SafeArg.of("datasetId", datasetId), 
+                UnsafeArg.of("sensitiveInfo", 42));
+    }
+}
+```
+
+The `HttpRemotingJerseyFeature` installs exception mappers for `ServiceException` The exception mapper sets the response
```
Comment (gracew) [inline readme.md]: period after ServiceException

File: readme.md [hunk 6]
```diff
@@ -227,28 +233,67 @@ Jackson `ObjectMapper` with `GuavaModule`, `ShimJdk7Module` (same as Jackson’s
...
+media type to `application/json` and returns as response body a JSON representation of a `SerializableError` capturing
+the error code, error name, and error parameters. The resulting JSON response is:
 
 ```json
 {
-  "message": "A string explaning the error",
-  "exceptionClass": "applicationSpecificErrorName",
-  "stackTrace": [ {"methodName":"...","fileName":"...","lineNumber":...,"className":"...","nativeMethod":false, {...} ]
+  "errorCode": "INVALID_ARGUMENT",
+  "errorName": "DatasetNotFound",
```
Comment (gracew) [inline readme.md]: namespace here as well

File: readme.md [hunk 7]
```diff
@@ -227,28 +233,67 @@ Jackson `ObjectMapper` with `GuavaModule`, `ShimJdk7Module` (same as Jackson’s
...
-traces.
+Both JaxRsClient and Retrofit2Client intercept non-successful HTTP responses and throw a `RemoteException` wrapping the
+deserialized server-side `SerializableError`. The error codes and names of the `ServiceException` and
+`SerializableError` are defined by the service API, and clients should handle errors based on the error code and name:
+
+```
+try {
+    service.someMethod();
+catch (RemoteExcetion e) {
+    if (e.getError().errorName().equals("DatasetNotFound")) {
```
Comment (gracew) [inline readme.md]: namespace

File: readme.md (hunk 1 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 2 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 4 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 3 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 5 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 6 above)
Comment (uschi2000) [reply readme.md]: done

File: readme.md (hunk 7 above)
Comment (uschi2000) [reply readme.md]: done

**Pull request merged**